netty/src/main/java/io/grpc/transport/netty/ProtocolNegotiators.java

```diff
@@ -224,18 +244,24 @@ public void flush(ChannelHandlerContext ctx) {
 
     @Override
     public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Exception {
-      failBufferedAndClose(ctx);
+      fail(ctx, new Exception("Channel closed while performing protocol negotiation"));
     }
 
-    protected void failBufferedAndClose(ChannelHandlerContext ctx) {
-      if (bufferedWrites != null) {
-        Exception e = new Exception("Buffered write failed.");
-        while (!bufferedWrites.isEmpty()) {
-          ChannelWrite write = bufferedWrites.poll();
-          write.promise.setFailure(e);
+    protected final void fail(ChannelHandlerContext ctx, Throwable cause) {
+      if (failureCause == null) {
+        failureCause = cause;
+        if (bufferedWrites != null) {
+          Exception e = new Exception("Buffered write failed.");
```

Comment:
Use `cause` instead?

That would also allow passing in a Status Exception which would allow us to set a Status other than UNKNOWN.